components/prize_pool/commons/prize_pool.lua

```diff
@@ -0,0 +1,352 @@
+			local pointsData = mw.loadData('Module:Points/data')
+			return pointsData[input] or {title = 'Points'}
+		end,
+		row = 'points',
+		rowParse = function (prizePool, input, context, index)
+			return prizePool:_parseInteger(input)
+		end
+	},
+	[prizeTypes.FREETEXT] = {
+		header = 'freetext',
+		headerParse = function (prizePool, input, context, index)
+			return {title = input}
+		end,
+		row = 'freetext',
+		rowParse = function (prizePool, input, context, index)
+			return input
+		end
+	},
+}
+
+local specialData = {
+	[specialDataTypes.GROUPSCORE] = {
+		field = 'wdl',
+		parse = function (prizePool, input, context)
+			return input
+		end
+	},
+	[specialDataTypes.LASTVS] = {
+		field = 'lastvs',
+		parse = function (prizePool, input, context)
+			return prizePool:_parseOpponentArgs(input, context.date)
+		end
+	},
+	[specialDataTypes.LASTVSSCORE] = {
+		field = 'lastvsscore',
+		parse = function (prizePool, input, context)
+			local scores = Table.mapValues(mw.text.split(input, '-'), tonumber)
+			return {score = scores[1], vsscore = scores[2]}
+		end
+	},
+}
+
+function PrizePool:init(args)
+	self.args = args
+	self.pagename = mw.title.getCurrentTitle().text
+	self.args.type = Json.parseIfString(self.args.type)
+	self.date = args.date or Variables.varDefaultMulti('tournament_enddate', 'tournament_edate', 'edate', TODAY)
+	if args.opponentLibrary then
+		Opponent = require('Module:'.. args.opponentLibrary)
+	end
+
+	if not self.args.type then
+		return error('Please provide a type!')
+	elseif type(self.args.type) ~= 'table' or not self.args.type.type then
+		return error('Could not parse type!')
+	elseif not Opponent.isType(self.args.type.type) then
+		return error('Not a valid type!')
+	end
+
+	self.opponentType = self.args.type.type
+
+	return self
+end
+
+function PrizePool:readInput(args)
+	self.options = self:_readStandardOptions(args)
+	self.prizes = self:_readPrizes(args)
+	self.placements = self:_readPlacements(args)
+
+	if self:_hasUsdPrizePool() then
+		self:setOption('showUSD', true)
+		self:addPrize('USD', 1)
+	end
+
+	return self
+end
+
+function PrizePool:create()
+	-- TODO
+	mw.logObject(self)
+	return ''
+end
+
+function PrizePool:_readStandardOptions(args)
+	local options = {}
+
+	self:setOption('autoUSD', Logic.nilOr(Logic.readBoolOrNil(args.autousd), true), options)
+	self:setOption('showUSD', false, options)
+	self:setOption('prizeSummary', Logic.nilOr(Logic.readBoolOrNil(args.prizesummary), true), options)
+	self:setOption('exchangeInfo', Logic.nilOr(Logic.readBoolOrNil(args.exchangeinfo), true), options)
+	self:setOption('storeSmw', Logic.nilOr(Logic.readBoolOrNil(args.storesmw), true), options)
+	self:setOption('storeLpdb', Logic.nilOr(Logic.readBoolOrNil(args.storelpdb), true), options)
+
+	return options
+end
+
+function PrizePool:_readPrizes(args)
+	local prizes = {}
+
+	for prizeEnum in pairs(prizeTypes) do
+		local prizeDatum = prizeData[prizeTypes[prizeEnum]]
+		local fieldName = prizeDatum.header
```

Comment:
Can `prizeDatum` be null?

Reply:
Only if there has been a serious error in the code. Every `prizeTypes` should have an entry in `prizeData`.

Reply:
Perhaps add one of swanky asserts here then 😉 ?

Reply:
Added an assert